File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
# 1. input
V, E = map(int, input().split())


# 2. Kruskal Algorithm
edges = []
for _ in range(E):
    A, B, C = map(int, input().split())
    edges.append((A, B, C))

edges.sort(key=lambda x: x[2])  # 가중치가 작은 것 부터 정렬


# 3. Union-Find
parent = [x for x in range(V+1)]

# 1) 특정 원소가 속한 집합 찾기
def find_parent(x):
    if parent[x] == x:
        return x
    parent[x] = find_parent(parent[x])
    return parent[x]


# 2) 두 원소가 속한 집합 합치기
def union_parent(a, b):
    a = find_parent(a)
    b = find_parent(b)

    if a < b:
        parent[b] = a
    else:
        parent[a] = b
        

# 3) 사이클 확인 -> 같은 부모를 공유하지 않으면 사이클 X
def same_parent(a, b):
    return find_parent(a) == find_parent(b)


# 4. result
# 사이클 X -> 추가
rlt = 0
for a, b, c in edges:
    if not same_parent(a, b):
        union_parent(a, b)  # 집합 합치기
        rlt += c

print(rlt)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.*;
import java.util.PriorityQueue;
import java.util.StringTokenizer;


// edge 정보 클래스
class Edge implements Comparable<Edge> {
    int a;
    int b;
    int weight;

    public Edge(int v1, int v2, int weight) {
        this.a = v1;
        this.b = v2;
        this.weight = weight;
    }

    @Override
    public int compareTo(Edge e) {
        return weight - e.weight;  // 정렬
    }
}


public class Main {

    private static int V, E;
    private static int[] parent;  //부모 정보
    static PriorityQueue<Edge> pq; // 간선 정보 저장

    // 2. Union-Find
    // 1) 노드의 부모 찾기
    private static int find_parent(int x) {

        if (parent[x] == x) return x;

        parent[x] = find_parent(parent[x]);
        return parent[x];
    }

    // 2) 두 원소의 집합 합치기 -> 부모 같게 만들기
    private static void union_parent(int a, int b) {
        a = find_parent(a);
        b = find_parent(b);

        // 작은 원소가 부모
        if (a < b) parent[b] = a;
        else parent[a] = b;
    }

    // 3) 사이클 확인 -> 같은 부모 X -> 사이클 X
    private static boolean same_parent(int a, int b) {
        return find_parent(a) == find_parent(b);
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        // 1. input
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());

        pq = new PriorityQueue<>();
        parent = new int[V + 1];

        //1) 부모 노드
        for (int i = 1; i <= V; i++) {
            parent[i] = i;
        }

        //2) 간선 정보
        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());

            int v1 = Integer.parseInt(st.nextToken());
            int v2 = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());

            pq.offer(new Edge(v1, v2, weight));
        }


        // 3. result
        int rlt = 0;
        while (!pq.isEmpty()) {
            Edge edge = pq.poll();

            int a = edge.a;
            int b = edge.b;

            if (!same_parent(a, b)) {
                union_parent(a, b);
                rlt += edge.weight;
            }
        }
        System.out.println(rlt);
    }
}